llvm/lib/CodeGen/AtomicExpandPass.cpp

```diff
@@ -2066,9 +2066,23 @@ bool AtomicExpandImpl::expandAtomicOpToLibcall(
     I->replaceAllUsesWith(V);
   } else if (HasResult) {
     Value *V;
-    if (UseSizedLibcall)
-      V = Builder.CreateBitOrPointerCast(Result, I->getType());
-    else {
+    if (UseSizedLibcall) {
+      // Add bitcasts from Result's scalar type to I's <n x ptr> vector type
+      if (I->getType()->getScalarType()->isPointerTy() &&
+          I->getType()->isVectorTy() && !Result->getType()->isVectorTy()) {
+        unsigned AS =
+            cast<PointerType>(I->getType()->getScalarType())->getAddressSpace();
+        ElementCount EC = cast<VectorType>(I->getType())->getElementCount();
+        Value *BC = Builder.CreateBitCast(
+            Result,
+            VectorType::get(IntegerType::get(Ctx, DL.getPointerSizeInBits(AS)),
```

Comment:
Use DL.getIntPtrType. You can also avoid all of the vector logic with Type::getWithNewType